File: src/models/Listing.js
export class Listing {
  constructor({
    id = crypto.randomUUID(),
    title = '',
    category = '',
    type = 'offer',  // 'offer' or 'request'
    description = '',
    area = '',
    contactPreference = 'chat',  // 'chat', 'call', 'whatsapp', 'in-person'
    status = 'active',  // 'active', 'saved', 'contacted', 'closed'
    neighborhoodId = 'bandra-west',
    createdAt = new Date().toISOString(),
    updatedAt = new Date().toISOString(),
    aiGenerated = false,  // whether description was AI-assisted
  } = {}) {
    this.id = id;
    this.title = this._stripHtml(title);
    this.category = category;
    this.type = type;
    this.description = this._stripHtml(description);
    this.area = this._stripHtml(area);
    this.contactPreference = contactPreference;
    this.status = status;
    this.neighborhoodId = neighborhoodId;
    this.createdAt = createdAt;
    this.updatedAt = updatedAt;
    this.aiGenerated = aiGenerated;
  }

  _stripHtml(str) {
    if (typeof str !== 'string') return str;
    return str.replace(/<[^>]*>?/gm, '').trim();
  }

  validate() {
    const errors = [];
    
    // Title
    if (!this.title) {
      errors.push('Title is required');
    } else if (this.title.length < 3) {
      errors.push('Title must be at least 3 characters');
    } else if (this.title.length > 100) {
      errors.push('Title must be at most 100 characters');
    }

    // Category
    if (!this.category) {
      errors.push('Category is required');
    }

    // Description
    if (!this.description) {
      errors.push('Description is required');
    } else if (this.description.length < 10) {
      errors.push('Description must be at least 10 characters');
    } else if (this.description.length > 1000) {
      errors.push('Description must be at most 1000 characters');
    }

    // Area
    if (!this.area) {
      errors.push('Area is required');
    } else if (/\d+/.test(this.area) && /(block|flat|apartment|apt|house|bldg|building|road|street)/i.test(this.area)) {
       // Rough heuristic to reject exact addresses
       errors.push('Area should be a general neighborhood area, not an exact address');
    }

    return {
      valid: errors.length === 0,
      errors
    };
  }

  toJSON() {
    return {
      id: this.id,
      title: this.title,
      category: this.category,
      type: this.type,
      description: this.description,
      area: this.area,
      contactPreference: this.contactPreference,
      status: this.status,
      neighborhoodId: this.neighborhoodId,
      createdAt: this.createdAt,
      updatedAt: this.updatedAt,
      aiGenerated: this.aiGenerated,
    };
  }

  static fromJSON(obj) {
    return new Listing(obj);
  }
}

File: lib/domain/models/listing.dart
import '../../config/app_config.dart';

enum ListingType { offer, request }
enum ListingStatus { active, saved, contacted, closed }
enum ContactPreference { whatsapp, call, inAppNote }

class ValidationResult {
  final bool isValid;
  final Map<String, String> errors;

  ValidationResult({required this.isValid, required this.errors});
}

class Listing {
  final String id;
  final String title;
  final String categoryId;
  final ListingType type; // offer or request
  final String description;
  final String area; // coarse sub-locality (e.g., "Pali Hill")
  final ContactPreference contactPreference;
  final ListingStatus status;
  final String neighborhoodId;
  final DateTime createdAt;
  final DateTime updatedAt;
  final bool aiGenerated;
  final int expiryDays; // Original Feature: Closing soon self-expiry tracking
  final String? imageUrl; // Visual image for the help / listing provided

  Listing({
    required this.id,
    required this.title,
    required this.categoryId,
    this.type = ListingType.offer,
    required this.description,
    required this.area,
    this.contactPreference = ContactPreference.whatsapp,
    this.status = ListingStatus.active,
    this.neighborhoodId = AppConfig.defaultNeighborhoodName,
    DateTime? createdAt,
    DateTime? updatedAt,
    this.aiGenerated = false,
    this.expiryDays = 7,
    this.imageUrl,
  })  : createdAt = createdAt ?? DateTime.now(),
        updatedAt = updatedAt ?? DateTime.now();

  /// Strip basic HTML tags from user input for safety
  static String sanitize(String input) {
    return input.replaceAll(RegExp(r'<[^>]*>?'), '').trim();
  }

  /// Check if closing soon (expires within 48 hours)
  bool get isClosingSoon {
    if (status == ListingStatus.closed) return false;
    final expiryDate = createdAt.add(Duration(days: expiryDays));
    final hoursRemaining = expiryDate.difference(DateTime.now()).inHours;
    return hoursRemaining <= 48 && hoursRemaining > 0;
  }

  /// Calculate days remaining before self-closing
  int get daysRemaining {
    final expiryDate = createdAt.add(Duration(days: expiryDays));
    final diff = expiryDate.difference(DateTime.now()).inDays;
    return diff < 0 ? 0 : diff;
  }

  /// Validate inputs before saving
  ValidationResult validate() {
    final errors = <String, String>{};

    final cleanTitle = sanitize(title);
    if (cleanTitle.isEmpty) {
      errors['title'] = 'Title is required';
    } else if (cleanTitle.length < 3) {
      errors['title'] = 'Title must be at least 3 characters';
    } else if (cleanTitle.length > 100) {
      errors['title'] = 'Title must be 100 characters or fewer';
    }

    if (categoryId.isEmpty) {
      errors['category'] = 'Category is required';
    }

    final cleanDesc = sanitize(description);
    if (cleanDesc.isEmpty) {
      errors['description'] = 'Description is required';
    } else if (cleanDesc.length < 10) {
      errors['description'] = 'Description must be at least 10 characters';
    } else if (cleanDesc.length > 1000) {
      errors['description'] = 'Description must be 1000 characters or fewer';
    }

    if (area.isEmpty) {
      errors['area'] = 'Neighborhood area is required';
    } else if (RegExp(r'\d+').hasMatch(area) &&
        RegExp(r'(flat|apt|apartment|room|bldg|building|street|road|st|lane)', caseSensitive: false).hasMatch(area)) {
      errors['area'] = 'Use a general sub-locality (e.g. Pali Hill), not an exact address';
    }

    return ValidationResult(isValid: errors.isEmpty, errors: errors);
  }

  Listing copyWith({
    String? id,
    String? title,
    String? categoryId,
    ListingType? type,
    String? description,
    String? area,
    ContactPreference? contactPreference,
    ListingStatus? status,
    String? neighborhoodId,
    DateTime? createdAt,
    DateTime? updatedAt,
    bool? aiGenerated,
    int? expiryDays,
    String? imageUrl,
  }) {
    return Listing(
      id: id ?? this.id,
      title: title ?? this.title,
      categoryId: categoryId ?? this.categoryId,
      type: type ?? this.type,
      description: description ?? this.description,
      area: area ?? this.area,
      contactPreference: contactPreference ?? this.contactPreference,
      status: status ?? this.status,
      neighborhoodId: neighborhoodId ?? this.neighborhoodId,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? DateTime.now(),
      aiGenerated: aiGenerated ?? this.aiGenerated,
      expiryDays: expiryDays ?? this.expiryDays,
      imageUrl: imageUrl ?? this.imageUrl,
    );
  }

  Map<String, dynamic> toMap() {
    return {
      'id': id,
      'title': title,
      'categoryId': categoryId,
      'type': type.name,
      'description': description,
      'area': area,
      'contactPreference': contactPreference.name,
      'status': status.name,
      'neighborhoodId': neighborhoodId,
      'createdAt': createdAt.toIso8601String(),
      'updatedAt': updatedAt.toIso8601String(),
      'aiGenerated': aiGenerated,
      'expiryDays': expiryDays,
      'imageUrl': imageUrl,
    };
  }

  factory Listing.fromMap(Map<String, dynamic> map) {
    return Listing(
      id: map['id'] ?? '',
      title: map['title'] ?? '',
      categoryId: map['categoryId'] ?? 'other',
      type: ListingType.values.firstWhere(
        (e) => e.name == map['type'],
        orElse: () => ListingType.offer,
      ),
      description: map['description'] ?? '',
      area: map['area'] ?? 'Bandra West',
      contactPreference: ContactPreference.values.firstWhere(
        (e) => e.name == map['contactPreference'],
        orElse: () => ContactPreference.whatsapp,
      ),
      status: ListingStatus.values.firstWhere(
        (e) => e.name == map['status'],
        orElse: () => ListingStatus.active,
      ),
      neighborhoodId: map['neighborhoodId'] ?? AppConfig.defaultNeighborhoodName,
      createdAt: DateTime.tryParse(map['createdAt'] ?? '') ?? DateTime.now(),
      updatedAt: DateTime.tryParse(map['updatedAt'] ?? '') ?? DateTime.now(),
      aiGenerated: map['aiGenerated'] ?? false,
      expiryDays: map['expiryDays'] ?? 7,
      imageUrl: map['imageUrl'],
    );
  }
}
